// @pollyjs/persister-fs/dist/cjs/pollyjs-persister-fs.js
/**
* @pollyjs/persister-fs v2.6.3
*
* https://github.com/netflix/pollyjs
*
* Released under the Apache-2.0 License.
*/
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var Persister = _interopDefault(require('@pollyjs/persister'));
var nodeServer = require('@pollyjs/node-server');

const {
  parse
} = JSON;
class FSPersister extends Persister {
  constructor() {
    super(...arguments);
    this.api = new nodeServer.API(this.options);
  }

  static get name() {
    return 'fs';
  }

  get defaultOptions() {
    return {
      recordingsDir: nodeServer.Defaults.recordingsDir
    };
  }

  findRecording(recordingId) {
    return this.api.getRecording(recordingId).body || null;
  }

  saveRecording(recordingId, data) {
    /*
      Pass the data through the base persister's stringify method so
      the output will be consistent with the rest of the persisters.
    */
    this.api.saveRecording(recordingId, parse(this.stringify(data)));
  }

  deleteRecording(recordingId) {
    this.api.deleteRecording(recordingId);
  }

}

module.exports = FSPersister;
